designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,372 @@
+
+### Disruption Behavior
+The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:
+
+AllowedDisruptions = minNodeCount - unhealthyNodes - totalDisruptingNodes.
+
+When calculating by reason, three potential equations emerge:
+1. BucketedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes[reason]
+2. AllowedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes
+3. minimumDisruptionAllowed = min(minNodeCount[reason], minNodeCount[default]) - unhealthyNodes - totalDisruptingNodes
+
+### Scenarios
+We will go through scenarios that users are expecting to face to drive home if the api can satisfy the users requirements for the end users of these features
+- Block Drift, but allow Expiration
+- Fully Block Drift && Expiration, but allow consolidation to occur 
+- Block Drift && Expiration, but allow emptiness 
+- Limit Drift && Expirationn while allowing agressive but not unbounded consolidationn
+- Allow Cost Saving operations only on weekends. 
+
+
+#### Block Drift and allow Expiration 
+Customers who roll out aggressive node image upgrades don't want to roll the nodes due to drift, but want to use node expiration to force users onto compliant node images eventually. The scenario here is an ML Training workload. Ideally we want to let it bake as long as possible, but not at the cost of compromising the security of the cluster. 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty 
+    expireAfter: 7d 
+    budgets:
+    - nodes: 0
+      reasons: [Drifted] 
+    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
+```
+
+We could also allow a semantic like this
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty 
+    expireAfter: 7d 
+    budgets:
+    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
+      reasons: [Expired, Empty]
+    - nodes: 0
+```
+So in this case, we would take the minimum node count, and it would block the 100% expiration budget of Expired and Empty. 
+But this semantic would not work for equation 3. Equation three will take the minimum between minNodeCount[reason] and the minNodeCount[default/unspecifed].  
+
+#### Fully Blocking Drift && Expiration while allowing consolidation
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: Never 
+    budgets:
+    - nodes: 0
+      reasons: [Drifted] 
```

Comment:
nit: need to add `Expired` to the array too?

Reply:
Ah, nvm I see:
![image](https://github.com/kubernetes-sigs/karpenter/assets/33269602/c00c87a8-2cc1-4ece-a3bf-1549c971b905)